feat!: Orchestration v2

ZhongpinWang opened the pull request:
## Context

Closes SAP/ai-sdk-js-backlog#362.

- [x] I know which base branch I chose for this PR, as the default branch is `v1-main` now, which is not for v2 development.
- [x] If my change will be merged into the `main` branch (for v2), I've updated (v2-Upgrade-Guide.md)[./v2-Upgrade-Guide.md] in case my change has any implications for users updating to SDK v2

## What this PR does and why it is needed

Findings:

1. AI Launchpad still generates orchestration v1 config. I would suggest we first not support reading from JSON until they update the config. I would avoid mapping this v1 config to v2, which is error-prone and beats the point of having a fresh v2 client.
2. `llm` and `templating` modules are merged into `prompt_templating` now in the config. However, the module results now aka. `intermediate_results` remain untouched. So still two separated modules in the response.
3. `stream: true` is removed from the top level of the config, we need to set `enabled: true` in global streaming options.

**What works now**

EVERYTHING! 
Orchestration v2 has replaced v1.
LangChain also uses v2.
All tests ✅

deekshas8 commented on packages/orchestration-v2/package.json:
```diff
@@ -0,0 +1,43 @@
+    "orchestration",
+    "llm-access"
+  ],
+  "type": "module",
```
[req] Set `private:true` here so we don't accidentally release it to npm with `next`

deekshas8 commented on packages/orchestration/src/orchestration-client.test.ts:
```diff
@@ -1,924 +1,926 @@
-    });
-  });
-});
+// import nock from 'nock';
```
[q] I don't get why we're commenting these tests out? New test files will be creating in the new package anyway right? Later we just delete everything in this folder.

deekshas8 reviewed (commented):
Is there anything specific that one should look at? The current state looks fine to me, given we haven't touched on the wrappers yet. Or am I missing something?

ZhongpinWang commented on packages/langchain/src/orchestration/client.test.ts:
```diff
@@ -1,494 +1,498 @@
-import { constructCompletionPostRequest } from '@sap-ai-sdk/orchestration/internal.js';
-import { jest } from '@jest/globals';
-import nock from 'nock';
-import { type AIMessageChunk } from '@langchain/core/messages';
```
Same, they use mock response from the `test-util` folder and we need to adjust the tests later.

ZhongpinWang commented on packages/orchestration-v2/src/internal.ts:
```diff
@@ -0,0 +1,4 @@
+export * from './orchestration-client.js';
+export * from './util/index.js';
+export * from './orchestration-types.js';
+export * from './orchestration-response.js';
```
I was simply not that far as I feel if urgent we can still do this in a follow-up. But already done now. 👍

ZhongpinWang commented:
> Is there anything specific that one should look at? The current state looks fine to me, given we haven't touched on the wrappers yet. Or am I missing something?

I think the changes are mostly straightforward, some mapping logic might be dangerous though, e.g., optional `template` property, optional `prompt` property etc.

For exporting items, if there is anything more needed to be exported publicly, we can added them later.

ZhongpinWang commented:
# Summary of SAP/ai-sdk-js Pull Request #924: "feat!: Orchestration v2"

*Generated by LLM*

## Overview
This PR introduces **Orchestration v2**, a major breaking change that completely restructures the SAP AI SDK's orchestration API. The changes span 78 files with 2,567 additions and 2,572 deletions across both the core orchestration package and LangChain integration.

## Key Structural Changes

### 1. Module Consolidation
**Major Change**: Merged `llm` and `templating` modules into a single `promptTemplating` module
- **Before**: Separate `llm` and `templating` configurations
- **After**: Unified `promptTemplating` module containing both model and prompt configuration

### 2. Configuration Structure Updates
**Model Configuration**:
- `llm.model_name` → `promptTemplating.model.name`
- `llm.model_params` → `promptTemplating.model.params`
- `llm.model_version` → `promptTemplating.model.version`

**Template Configuration**:
- `templating.template` → `promptTemplating.prompt.template`

### 3. Parameter Name Changes
**Request Parameters**:
- `inputParams` → `placeholderValues` (in client method calls)
- `input_params` → `placeholder_values` (in request schema)
- `orchestration_config` → `config` (in request schema)

**Response Properties**:
- `orchestration_result` → `final_result`
- `module_results` → `intermediate_results`

### 4. Streaming Configuration
**Before**: Top-level `stream: true` property
**After**: `streamOptions.enabled: true` with module-specific options under `promptTemplating`

### 5. Grounding Configuration
**Before**: Separate `input_params` and `output_param` properties
**After**: Nested `placeholders` object with `input` and `output` properties

### 6. Azure Content Filter Updates
**Property Name Changes** (PascalCase → snake_case):
- `Hate` → `hate`
- `SelfHarm` → `self_harm`
- `Sexual` → `sexual`
- `Violence` → `violence`
- `PromptShield` → `prompt_shield`

**Function Removal**:
- Removed deprecated `buildAzureContentFilter()`
- Use `buildAzureContentSafetyFilter()` instead

## LangChain Integration Changes

### Configuration Updates
- Updated `LangChainOrchestrationModuleConfig` to use `promptTemplating` structure
- Removed deprecated `LangchainOrchestrationModuleConfig` type alias

### Response Property Updates
- Message `additional_kwargs.module_results` → `additional_kwargs.intermediate_results`
- Updated `OrchestrationMessage` and `OrchestrationMessageChunk` classes

## API Schema Changes

### Request/Response Structure
- Updated endpoint from `/completion` to `/v2/completion`
- New error response structure with nested `error` object
- Added new schema types: `Error`, `ErrorStreaming`, `LLMModelDetails`
- Removed: `LlmModuleConfig`, `InputTranslationModuleConfig`, `OutputTranslationModuleConfig`

### Type System Updates
- Consolidated translation modules into single `TranslationModuleConfig`
- Updated import paths to use `/internal.js` for internal types
- Enhanced type safety with `Xor` utility types

## Testing Updates
- Updated all test snapshots to reflect new response structure
- Modified test configurations to use new `promptTemplating` structure
- Updated mock responses and test data

## Documentation
- Comprehensive V2 Upgrade Guide with before/after examples
- Detailed changeset files documenting all breaking changes
- Updated sample code and configuration examples

## Impact Assessment
This is a **complete API redesign** that:
- **Breaks all existing v1 integrations**
- **Requires full migration** for users upgrading to v2
- **Modernizes the API structure** for better consistency
- **Maintains full functionality** while improving developer experience
- **Aligns with updated service APIs** from the backend

The PR represents a significant architectural improvement that consolidates fragmented configuration patterns into a more cohesive and intuitive API design.

tomfrenken commented on packages/langchain/src/orchestration/util.ts:
```diff
@@ -341,14 +333,14 @@ export function isToolDefinitionLike(
   chunk: OrchestrationStreamChunkResponse
 ): AIMessageChunk {
-  const choice = chunk.data.orchestration_result?.choices[0];
+  const choice: any = chunk.data.final_result?.choices[0];
```
[q] Why any?

tomfrenken commented on packages/orchestration/src/client/api/schema/template.ts:
```diff
@@ -14,7 +14,7 @@ import type { ChatCompletionTool } from './chat-completion-tool.js';
  *     {
  *       "role": "user",
- *       "content": "How can the features of AI in SAP BTP specifically {{?groundingOutput}}, be applied to {{?inputContext}}"
+ *       "content": "How can the features of AI in SAP BTP specifically {{'{{?groundingOutput}}'}}, be applied to {{'{{?inputContext}}'}}"
```
Generated, but this looks so cursed 💀 .

tomfrenken commented on packages/orchestration/package.json:
```diff
@@ -34,7 +34,7 @@
     "@sap-ai-sdk/prompt-registry": "workspace:^",
     "@sap-cloud-sdk/util": "^4.1.1",
-    "yaml": "^2.8.1"
+    "yaml": "^2.8.0"
```
[q] Why downgrade?

tomfrenken reviewed (approved):
LGTM, I already dread the merge conflcits with my PR 😢
